diff --git a/compressing_classifiers_and_MLPs/imagenet.jl b/compressing_classifiers_and_MLPs/imagenet.jl
index 50af284..c23cc89 100644
--- a/compressing_classifiers_and_MLPs/imagenet.jl
+++ b/compressing_classifiers_and_MLPs/imagenet.jl
@@ -1,8 +1,10 @@
-using Pkg; Pkg.activate(".")
+using Pkg; Pkg.activate("."); 
+# Pkg.resolve(); Pkg.instantiate()
 using Revise
 
 using CUDA, TOML, ArgParse, Suppressor, Optimisers, ParameterSchedulers
 using Lux: gpu_device
+using TOML: parsefile
 
 using CompressingClassifiersMLPs.TrainingArguments: TrainArgs
 using CompressingClassifiersMLPs.OptimizationProcedures: PMMP_procedure,
@@ -17,8 +19,8 @@ using CompressingClassifiersMLPs.OptimizationProcedures: PMMP_procedure,
     alexnet
 using CompressingClassifiersMLPs.DatasetsModels: MNIST_data, 
     CIFAR_data, 
-    imagenet_data_function, 
-    toy_imagenet_data_function
+    imagenet_data, 
+    toy_imagenet_data
 using CompressingClassifiersMLPs.BatchRun: do_batch_run, 
     get_sub_batch,
     single_run_routine_classifier,
@@ -31,7 +33,7 @@ args = TrainArgs{Float32}()
 
 # Load configuration
 @assert isfile("config.toml") "File `config.toml` does not exist or script run from wrong path."
-cfg = TOML.parsefile("config.toml")
+cfg = parsefile("config.toml")
 @assert haskey(cfg, "paths") "config file should have [paths] section"
 @assert haskey(cfg["paths"], "path_to_db")
 @assert haskey(cfg["paths"], "imagenet_path")
@@ -144,9 +146,11 @@ append!(batch, DRR_runs)
 # append!(batch, FPP_runs)
 
 # Fixed arguments for all runs
-args.dataset = toy_imagenet_data_function
+imagenet_data_function = trainbatchsize -> imagenet_data(imagenet_path, trainbatchsize, trainbatchsize, 224; dev=gpu_device())
+toy_imagenet_data_function = trainbatchsize -> toy_imagenet_data(imagenet_path, trainbatchsize, trainbatchsize, 224; dev=gpu_device())
+args.dataset = imagenet_data_function # toy_imagenet_data_function # imagenet_data_function
 
-args.architecture = toy_resnet
+args.architecture = resnet50 #toy_resnet # resnet50
 args.delete_neurons = false
 args.layerwise_pruning = false
 args.smoothing_window = 5
@@ -186,12 +190,12 @@ args.schedule = Step(
 )
 
 args.multiply_mask_after_each_batch = true
-args.debug = true
+args.debug = false # false
 
 args.use_checkpoints = true
 args.checkpoint_dir = joinpath(path_to_db, experiment_name, "checkpoints")
 args.checkpoint_frequency = 1
-args.max_runtime_seconds = 3600 * 23.5 / 5 # /5 for dev
+args.max_runtime_seconds = 3600 * 23.5
 
 break_if_one_run_errors = true
 
diff --git a/compressing_classifiers_and_MLPs/src/BatchRun/single_run_routine_classifier.jl b/compressing_classifiers_and_MLPs/src/BatchRun/single_run_routine_classifier.jl
index 324990e..699dd33 100644
--- a/compressing_classifiers_and_MLPs/src/BatchRun/single_run_routine_classifier.jl
+++ b/compressing_classifiers_and_MLPs/src/BatchRun/single_run_routine_classifier.jl
@@ -46,7 +46,7 @@ function single_run_routine_classifier(
 
     model_seed = args.seed + 42; loss_fctn = logitcrossentropy;
     
-    start_time = time()
+    start_time = now()
 
     println("Start training for $run_id with architecture '$(args.architecture)', dataset '$(args.dataset)' and optimization procedure '$(args.optimization_procedure)'")
 
@@ -75,6 +75,8 @@ function single_run_routine_classifier(
         tstate = generate_tstate(model, model_seed, args.optimizer(args.lr); dev=args.dev)
     end
 
+    @assert tstate != nothing
+
     tstate, logs, loss_fctn, checkpoint = args.optimization_procedure(train_set, validation_set, test_set, tstate, loss_fctn, args, checkpoint)
 
     # save results
diff --git a/compressing_classifiers_and_MLPs/src/Checkpointer.jl b/compressing_classifiers_and_MLPs/src/Checkpointer.jl
index f2ff3a2..eb0359c 100644
--- a/compressing_classifiers_and_MLPs/src/Checkpointer.jl
+++ b/compressing_classifiers_and_MLPs/src/Checkpointer.jl
@@ -48,6 +48,13 @@ function maybe_save_checkpoint(checkpoint::CheckpointManager)
     end
 end
 function save_checkpoint(metadata::CheckpointMetadata, content::CheckpointContent)
+
+    for field in fieldnames(CheckpointContent)
+        if getfield(content, field) == nothing 
+            @warn "CheckpointContent field `$(field)` is `nothing`."
+        end
+    end
+
     mkpath(metadata.path)
     path = joinpath(metadata.path, metadata.checkpoint_id*".jld2")
     jldopen(path, "w") do f
@@ -135,6 +142,12 @@ function load_checkpoint(filepath::String, args::AbstractTrainArgs)::Tuple{Check
     content = jldopen(filepath, "r") do f
         f["content"]
     end
+    
+    if content.best_tstate == nothing && content.tstate != nothing
+        @warn "Checkpoint loaded with a `nothing` best_tstate. Substituting for tstate."
+        content.best_tstate = deepcopy(content.tstate)
+    end
+    
     content.tstate |> args.dev
     content.best_tstate |> args.dev
 
diff --git a/compressing_classifiers_and_MLPs/src/DatasetsModels/DatasetsModels.jl b/compressing_classifiers_and_MLPs/src/DatasetsModels/DatasetsModels.jl
index 4e3a4f5..ea0ff7c 100644
--- a/compressing_classifiers_and_MLPs/src/DatasetsModels/DatasetsModels.jl
+++ b/compressing_classifiers_and_MLPs/src/DatasetsModels/DatasetsModels.jl
@@ -16,9 +16,9 @@ module DatasetsModels
     export CIFAR_data
 
     include("imagenet_dataset.jl")
-    export imagenet_data, magenet_data_function
+    export imagenet_data
 
     include("toy_imagenet_dataset.jl")
-    export imagenet_data, toy_imagenet_data_function
+    export toy_imagenet_data
 
 end # module
\ No newline at end of file
diff --git a/compressing_classifiers_and_MLPs/src/DatasetsModels/imagenet_dataset.jl b/compressing_classifiers_and_MLPs/src/DatasetsModels/imagenet_dataset.jl
index a0a088d..2ae4be2 100644
--- a/compressing_classifiers_and_MLPs/src/DatasetsModels/imagenet_dataset.jl
+++ b/compressing_classifiers_and_MLPs/src/DatasetsModels/imagenet_dataset.jl
@@ -140,5 +140,4 @@ function imagenet_data(base_path::String, train_batchsize, val_batchsize, image_
     test_set = nothing
     return train_set, val_set, test_set
 end
-imagenet_data_function = trainbatchsize -> imagenet_data(imagenet_path, trainbatchsize, trainbatchsize, 224; dev=gpu_device())
 
diff --git a/compressing_classifiers_and_MLPs/src/DatasetsModels/toy_imagenet_dataset.jl b/compressing_classifiers_and_MLPs/src/DatasetsModels/toy_imagenet_dataset.jl
index 479c4e6..b2f8a69 100644
--- a/compressing_classifiers_and_MLPs/src/DatasetsModels/toy_imagenet_dataset.jl
+++ b/compressing_classifiers_and_MLPs/src/DatasetsModels/toy_imagenet_dataset.jl
@@ -76,4 +76,3 @@ function toy_imagenet_data(
 
     return train_dl, val_dl, nothing
 end
-toy_imagenet_data_function = trainbatchsize -> toy_imagenet_data(imagenet_path, trainbatchsize, trainbatchsize, 224; dev=gpu_device())
\ No newline at end of file
diff --git a/compressing_classifiers_and_MLPs/src/OptimizationProcedures/DRR_procedure.jl b/compressing_classifiers_and_MLPs/src/OptimizationProcedures/DRR_procedure.jl
index 21e6de2..8e45903 100644
--- a/compressing_classifiers_and_MLPs/src/OptimizationProcedures/DRR_procedure.jl
+++ b/compressing_classifiers_and_MLPs/src/OptimizationProcedures/DRR_procedure.jl
@@ -63,6 +63,8 @@ function DRR_procedure(
         loss_fun = DRR(model_param_number, layernumber_model; alpha=args.α, beta=args.β, rho=args.ρ, loss_f=loss_fctn, fun1 = fun1)
     end
 
+    @assert tstate != nothing
+
     return procedure(train_set, validation_set, test_set, tstate, loss_fun, args, checkpoint)
 end
 
diff --git a/compressing_classifiers_and_MLPs/src/OptimizationProcedures/HelperFunctions/lux_training.jl b/compressing_classifiers_and_MLPs/src/OptimizationProcedures/HelperFunctions/lux_training.jl
index 3834a2b..d874714 100644
--- a/compressing_classifiers_and_MLPs/src/OptimizationProcedures/HelperFunctions/lux_training.jl
+++ b/compressing_classifiers_and_MLPs/src/OptimizationProcedures/HelperFunctions/lux_training.jl
@@ -142,8 +142,8 @@ function lux_training!(
     num_batches = args.dtype(length(train_set))
 
     # Checkpoint loading and initialization of variables
-    if checkpoint.do_checkpointing && checkpoint_enabled && checkpoint.content.epoch > 0
-        println("Starting/Resuming from epoch $(checkpoint.content.epoch)")
+    if checkpoint.do_checkpointing && checkpoint_enabled && checkpoint.content.epoch > 1
+        println("Resuming from epoch $(checkpoint.content.epoch)")
         start_epoch_offset = checkpoint.content.epoch - 1
         prev_val_loss = checkpoint.content.prev_val_loss
         total_time_start = checkpoint.metadata.start_time
@@ -167,7 +167,7 @@ function lux_training!(
     for epoch in (start_epoch_offset+1):max_epochs
         # Check for timeout before starting epoch
         if checkpoint.do_checkpointing && checkpoint_enabled && should_stop_for_timeout(checkpoint.metadata)
-            println("Maximum runtime approaching. Saving checkpoint and exiting...")
+            println("Saving checkpoint")
             update_checkpoint_state!(
                 checkpoint;
                 args=args,
@@ -181,6 +181,9 @@ function lux_training!(
                 status=:available
             )
             maybe_save_checkpoint(checkpoint)
+        end
+        if heckpoint.do_checkpointing && checkpoint_enabled && should_stop_for_timeout(checkpoint.metadata)
+            println("Maximum runtime approaching. Saving checkpoint and exiting...")
             error("MaxRuntimeReached")
         end
 
@@ -198,10 +201,9 @@ function lux_training!(
         for (i, batch) in enumerate(train_set)
             batch_time = time()
             if args.debug 
-                println("▶ Epoch $epoch – batch $i  start update-step")
+                println("▶ Epoch $epoch – batch $i  update-step finished after $(time() - batch_time) s")
             end
             tstate, loss, stats = update_state!(vjp, loss_fun, batch, tstate)
-            println("▶ Epoch $epoch – batch $i  update-step finished after $(time() - batch_time) s")
             if haskey(tstate.states, :mask) && args.multiply_mask_after_each_batch
                 recursively_multiply!(tstate.parameters.p, tstate.states.mask)
             end
@@ -219,11 +221,13 @@ function lux_training!(
         push!(args.logs["epochs"], start_epoch+epoch)
         push!(args.logs["train_loss"], epoch_loss)
         push!(args.logs["epoch_execution_time"], epoch_end_time - epoch_start_time)
+
+        println("▶ Epoch $epoch finished in $(time() - batch_time) s")
         
-        if args.debug
-            metrics_time = time()
-            println("▶ Epoch $epoch Start rest of training loop evaluations")
-        end
+        # if args.debug
+            # metrics_time = time()
+            # println("▶ Epoch $epoch Start rest of training loop evaluations")
+        # end
 
         if haskey(tstate.states, :mask)
             l0_mask= recursive_sum(tstate.states.mask, args.dtype(0))
@@ -231,6 +235,7 @@ function lux_training!(
         end
         if args.log_val_loss
             epoch_val_loss = zero(args.dtype)
+            t = time()
             for (i, val_batch) in enumerate(validation_set)
                 epoch_val_loss += loss_fun(tstate.model, tstate.parameters, Lux.testmode(tstate.states), val_batch)[1]
                 if args.debug && i > 5
@@ -242,14 +247,17 @@ function lux_training!(
             if args.verbose 
                 @printf "\rEpoch: %5d \t Train %.4g \t Val_loss: %.4g \t Time: %.2f \t" epoch epoch_loss epoch_val_loss (epoch_end_time - epoch_start_time)
             end
+            println("▶ Validiation loss evaluated in $(time()-t)s")
 
             if !isnothing(test_set)
+                t = time()
                 epoch_test_loss = zero(args.dtype)
                 for test_batch in test_set
                     epoch_test_loss += loss_fun(tstate.model, tstate.parameters, Lux.testmode(tstate.states), test_batch)[1]
                 end
                 epoch_test_loss /= length(test_set)
                 push!(args.logs["test_loss"], (start_epoch+epoch, epoch_test_loss))
+                println("▶ Test loss evaluated in $(time()-t)s")
             end            
         end
         if args.verbose && !args.log_val_loss
@@ -259,11 +267,13 @@ function lux_training!(
             push!(args.logs["sigmas"], sum(tstate.parameters.sigma))
         end
         if loss_fun.loss_f == logitcrossentropy
-            # this line is super slow for some reason
+            t = time()
             push!(args.logs["validation_accuracy"], accuracy(tstate, validation_set, debug=args.debug))
+            println("▶ Validation accuracy evaluated in $(time()-t)s")
         end
         
         # Pruning and Convergence check is done at most every prune_window epochs
+        time_pruning_and_convergence = time()
         if epoch % args.prune_window == 0 || epoch == max_epochs
 
             if converge_val_loss && args.log_val_loss
@@ -330,9 +340,10 @@ function lux_training!(
                 end
             end
         end
-        if args.debug
-            println("▶ Epoch $epoch - other evaluations took $(time() - metrics_time) s")
-        end
+        println("▶ Pruning and convergence check took $(time()-time_pruning_and_convergence)s")
+        # if args.debug
+            # println("▶ Epoch $epoch - other evaluations took $(time() - metrics_time) s")
+        # end
     end
     total_time_end = time()
     args.logs["total_time"] += total_time_end - total_time_start
@@ -343,6 +354,8 @@ function lux_training!(
         push!(args.logs["converged_at"], max_epochs)
     end
 
+    @assert tstate != nothing
+
     # take the tstate closest to where val_loss was smallest:
     if converge_val_loss && args.log_val_loss
         comparison_losses = args.logs["val_loss"][args.logs["turning_points_val_loss"][start_turn_point+1:end]]
@@ -372,6 +385,8 @@ function lux_training!(
         maybe_save_checkpoint(checkpoint)
     end
 
+    @assert return_tstate != nothing
+
     return return_tstate, args.logs, loss_fun, checkpoint
 end
 
@@ -387,6 +402,12 @@ function update_checkpoint_state!(
     convergence_triggered,
     status::Symbol
 )
+
+    if best_tstate === nothing
+        @warn "best_tstate is `nothing` when saving checkpoint – using current tstate as fallback."
+        best_tstate = tstate
+    end
+
     checkpoint.metadata.status = status
     checkpoint.metadata.last_updated = time()
 
@@ -398,3 +419,4 @@ function update_checkpoint_state!(
     checkpoint.content.best_tstate = best_tstate
     checkpoint.content.convergence_triggered = convergence_triggered
 end
+
diff --git a/compressing_classifiers_and_MLPs/src/OptimizationProcedures/HelperFunctions/tamade.jl b/compressing_classifiers_and_MLPs/src/OptimizationProcedures/HelperFunctions/tamade.jl
index f59fe36..e05ebc1 100644
--- a/compressing_classifiers_and_MLPs/src/OptimizationProcedures/HelperFunctions/tamade.jl
+++ b/compressing_classifiers_and_MLPs/src/OptimizationProcedures/HelperFunctions/tamade.jl
@@ -144,7 +144,18 @@ end
         - `random_gradient_pruning`: Boolean that determines whether random gradient pruning should be performed to eliminate spurious weights or not.
         - `final_epoch`: A Boolean that indicates whether training reached its final epoch or not. This is important for the PMMP procedure where pruning should only be performed in the last epoch to prevent interference with the PMMP mask.
 """
-function prune_and_shrink!(tstate, loss_fun, data, tolerance, binary_search_resolution=1e-7 ; dtype=Float32, dev=cpu_device(), delete_neurons=true, random_gradient_pruning=true, final_epoch=false)
+function prune_and_shrink!(
+    tstate::Lux.Training.TrainState, 
+    loss_fun, 
+    data, 
+    tolerance, 
+    binary_search_resolution::Number=1e-7 ; 
+    dtype=Float32, 
+    dev=cpu_device(), 
+    delete_neurons::Bool=true, 
+    random_gradient_pruning::Bool=true, 
+    final_epoch::Bool=false
+    )
     
     if haskey(tstate.states, :mask)
         recursively_set_to_zero!(tstate.parameters.p, tstate.states.mask)
